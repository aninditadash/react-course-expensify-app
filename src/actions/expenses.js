import { v4 as uuidv4 } from "uuid";

import database from "../firebase/firebase";
import { ref, set, push, get, remove, update } from "firebase/database";

// Action generators for EXPENSES

// Normal application flow in redux
// component calls action generator
// component dispatches object
// redux store changes
//

// saving data to Firebase ----->
// component calls action generator
// component dispatches function
// function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_EXPENSE
// export const addExpense = ({
//   description = "",
//   note = "",
//   amount = 0,
//   createdAt = 0
// } = {}) => ({
//   type: "ADD_EXPENSE",
//   expense: {
//     id: uuidv4(),
//     description,
//     note,
//     amount,
//     createdAt
//   }
// });

// ADD_EXPENSE
export const addExpense = (expense = {}) => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAppExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };
    const dataRef = push(ref(database, "expenses"));
    return set(dataRef, expense).then(() => {
      dispatch(
        addExpense({
          id: dataRef.key,
          ...expense
        })
      );
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id = "" } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id = "" }) => {
  return (dispatch) => {
    return remove(ref(database, `expenses/${id}`)).then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return update(ref(database, `expenses/${id}`), updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return (dispatch) => {
    const dataRef = ref(database, "expenses");
    return get(dataRef)
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      })
      .catch((e) => console.log("Error fetching data:: ", e));
  };
};
