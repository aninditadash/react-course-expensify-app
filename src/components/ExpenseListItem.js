import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <p>Description: {description}</p>
    <p>Amount: {amount}</p>
    <p>Created At: {createdAt}</p>

    <Link to={`/edit/${id}`}>Edit</Link>
  </div>
);

export default ExpenseListItem;
