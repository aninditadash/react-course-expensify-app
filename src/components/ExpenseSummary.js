import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";

export const ExpenseSummary = ({ expenseCount, expensesTotal }) => {
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          {expenseCount > 1 ? (
            <p className="page-header__title-bold">
              Viewing {expenseCount} expenses totalling{" "}
              <span>{numeral(expensesTotal).format("$0,0.00")}</span>
            </p>
          ) : (
            <p>
              Viewing {expenseCount} expense totalling{" "}
              <span>{numeral(expensesTotal).format("$0,0.00")}</span>
            </p>
          )}
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => ({
  expenseCount: selectExpenses(state.expenses, state.filters).length,
  expensesTotal: getExpensesTotal(selectExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpenseSummary);
