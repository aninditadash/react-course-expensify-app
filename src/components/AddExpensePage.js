import React from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";

import ExpenseForm from "./ExpenseForm";
import { startAppExpense } from "../actions/expenses";

export const AddExpensePage = (props) => {
  const navigate = useNavigate();

  const onSubmit = (expense) => {
    props.startAppExpense(expense);
    navigate("/");
  };

  return (
    <>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Add Expense</h1>
        </div>
      </div>
      <div className="content-container">
        <ExpenseForm onSubmit={onSubmit} />
      </div>
    </>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startAppExpense: (expense) => dispatch(startAppExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
