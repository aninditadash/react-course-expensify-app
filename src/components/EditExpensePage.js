import React from "react";
import { connect } from "react-redux";
import { useParams, useNavigate } from "react-router-dom";

import ExpenseForm from "./ExpenseForm";

import { editExpense, removeExpense } from "../actions/expenses";

export const EditExpensePage = (props) => {
  let { id } = useParams();
  const navigate = useNavigate();
  const selectedExpense = props.expenses.find((expense) => expense.id === id);

  const onSubmit = (expense) => {
    props.editExpense(id, expense);
    navigate("/");
  };

  const onRemove = () => {
    props.removeExpense({ id });
    navigate("/");
  };

  return (
    <div>
      <ExpenseForm expense={selectedExpense} onSubmit={onSubmit} />
      <button onClick={onRemove}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses
  };
};

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
