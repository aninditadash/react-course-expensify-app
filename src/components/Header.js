import React from "react";
import { connect } from "react-redux";
import { NavLink } from "react-router-dom";
import { startLogout } from "../actions/auth";

// import { history } from "../routers/AppRouter";

// import history from "../history";

export const Header = ({ startLogout }) => (
  <header>
    <h1>Expensify</h1>
    <NavLink
      to="/dashboard"
      className={({ isActive }) => (isActive ? "is-active" : undefined)}
    >
      Dashboard
    </NavLink>
    <NavLink
      to="/create"
      className={({ isActive }) => (isActive ? "is-active" : undefined)}
    >
      Create Expense
    </NavLink>
    <NavLink
      to="/edit"
      className={({ isActive }) => (isActive ? "is-active" : undefined)}
    >
      Edit Expense
    </NavLink>
    <NavLink
      to="/help"
      className={({ isActive }) => (isActive ? "is-active" : undefined)}
    >
      Help
    </NavLink>
    <button onClick={startLogout}>Logout</button>
  </header>
);

const mapStateToProps = (dispatch) => ({
  startLogout: () => {
    // history.push("/");
    return dispatch(startLogout());
  }
});

export default connect(undefined, mapStateToProps)(Header);
