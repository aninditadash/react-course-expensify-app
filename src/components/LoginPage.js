import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";
// import { useNavigate } from "react-router-dom";

// import history from "../history";
// import { store } from "../app";

// import { history } from "../routers/AppRouter";

export const LoginPage = ({ startLogin }) => {
  // const navigate = useNavigate();

  //   const onSubmit = (expense) => {
  //     props.startAppExpense(expense);
  //     navigate("/");
  //   };
  // debugger;

  // const login = () => {
  //   startLogin();
  //   // navigate("/dashboard");
  // };

  return (
    <div>
      Login Page
      <button onClick={startLogin}>Login</button>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
  // {
  //   // store.dispatch(startLogin()).then(() => {
  //   //   render();
  //   //   if (history.location.pathname === "/") {
  //   //     history.push("/dashboard");
  //   //   }
  //   // });
  //   // navigate("/dashboard");
  //   return .then(() => {
  //     // debugger;
  //     // navigate("/dashboard");
  //     // history.push("/dashboard");
  //   });
  // }
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
