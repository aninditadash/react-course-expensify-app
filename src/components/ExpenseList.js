import React from "react";
import { connect } from "react-redux";

import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="expense-list__header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="expense-list__body">
      {props.expenses.length === 0 ? (
        <div className="expense-list-item  expense-list__message">
          No expenses
        </div>
      ) : (
        props.expenses.map((expense) => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )}
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
