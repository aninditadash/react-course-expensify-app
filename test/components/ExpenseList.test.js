import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../src/components/ExpenseList";
import expenses from "../fixtures/expenses";

test("ExpenseList Component: Should render ExpenseList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("ExpenseList Component: Should render ExpenseList with empty message", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
