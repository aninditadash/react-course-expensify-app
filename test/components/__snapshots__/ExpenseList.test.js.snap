// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList Component: Should render ExpenseList with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`ExpenseList Component: Should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={4500}
    createdAt={0}
    description="Water bill"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={2500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note="This is the rent of the house"
  />
  <ExpenseListItem
    amount={5500}
    createdAt={345600000}
    description="Gas bill"
    id="3"
    key="3"
    note=""
  />
</div>
`;
